src/promptflow/tests/executor/unittests/contracts/test_flow.py

```diff
@@ -58,3 +71,373 @@ def test_node_condition_conflict(self):
                 Flow.deserialize(yaml.safe_load(fin))
         error_message = "Node 'test_node' can't have both skip and activate condition."
         assert str(e.value) == error_message, "Expected: {}, Actual: {}".format(error_message, str(e.value))
+
+    @pytest.mark.parametrize(
+        "flow_folder",
+        [
+            ("web_classification"),
+            ("flow_with_dict_input_with_variant"),
+        ],
+    )
+    def test_flow_serialize(self, flow_folder):
+        flow_yaml = get_yaml_file(flow_folder)
+        flow = Flow.from_yaml(flow_yaml)
+        serialized_flow = flow.serialize()
+        assert serialized_flow["name"] == flow.name
+        assert serialized_flow["nodes"] == [n.serialize() for n in flow.nodes]
+        assert serialized_flow["inputs"] == {name: i.serialize() for name, i in flow.inputs.items()}
+        assert serialized_flow["tools"] == [t.serialize() for t in flow.tools]
+
+        deserialize_flow = Flow.deserialize(serialized_flow)
+        assert deserialize_flow.name == flow.name
+        assert len(deserialize_flow.nodes) == len(flow.nodes)
+        assert deserialize_flow.inputs == flow.inputs
+        assert deserialize_flow.outputs == flow.outputs
+        assert deserialize_flow.tools == flow.tools
+
+    def test_import_requisites(self):
+        tool1 = Tool(name="tool1", type=ToolType.PYTHON, inputs={"name": InputDefinition(type=["int"])}, module="yaml")
+        tool2 = Tool(
+            name="tool2", type=ToolType.PYTHON, inputs={"name": InputDefinition(type=["int"])}, module="module"
+        )
+        node1 = Node(
+            name="node1", tool="tool1", inputs={"name": InputAssignment("value", InputValueType.LITERAL)}, module="yaml"
+        )
+        node2 = Node(
+            name="node2",
+            tool="tool2",
+            inputs={"name": InputAssignment("value", InputValueType.LITERAL)},
+            module="module",
+        )
+        tools = [tool1, tool2]
+        nodes = [node1, node2]
+
+        with pytest.raises(FailedToImportModule) as e:
+            Flow._import_requisites(tools, nodes)
+
+        assert str(e.value).startswith("Failed to import modules with error:")
+
+    def test_apply_default_node_variants(self):
+        flow_folder = "flow_with_dict_input_with_variant"
+        flow_yaml = get_yaml_file(flow_folder)
+        flow = Flow.from_yaml(flow_yaml)
+        # test when node.use_variants is True
+        flow._apply_default_node_variants()
+        assert flow.nodes[0].use_variants is False
+        assert flow.nodes[0].inputs == flow.node_variants["print_val"].variants["variant1"].node.inputs
+        # test when node.use_variants is False
+        flow = Flow.from_yaml(flow_yaml)
+        flow.nodes[0].use_variants = False
+        tmp_nodes = flow.nodes
+        flow._apply_default_node_variants()
+        assert flow.nodes == tmp_nodes
+
+    def test_apply_default_node_variant(self):
```

Comment:
Most of the cases are not unit test style. Unit tests should test only one method, and cover all the conditions (branches). 
The current code is most likely to be an integration test. It depends on external methods/data.

Let's discuss offline for the details tomorrow.

Reply:
For some functions must depend on external data such as get_connection_names, I choose to retain the e2etest in the current file. The other test cases are now changed to simple examples.

Reply:
Thanks. Sounds good to me.